prologin/matchmaker/matchmaker.py

```diff
@@ -0,0 +1,397 @@
+        """Wires the monitoring probes."""
+        matchmaker_queue_size.set_function(lambda: len(self.next_matches))
+        matchmaker_managed_champion_count.set_function(
+            lambda: len(self.champions)
+        )
+        matchmaker_latest_champion_age_seconds.set_function(
+            lambda: (
+                datetime.datetime.now()
+                - max(
+                    self.champions,
+                    key=lambda elt: elt.ts,
+                ).ts.date()
+            ).total_seconds()
+            if self.champions
+            else 0
+        )
+        matchmaker_managed_match_count.set_function(
+            lambda: sum(
+                len(matches) for matches in self.matches_with_champion.values()
+            )
+        )
+        matchmaker_match_limit_reached.set_function(self.match_limit_reached)
+
+    def run(self, *args, **kwargs):
+        logging.info(
+            "matchmaker listening on port %s",
+            self.config["matchmaker"]["port"],
+        )
+
+        # Start the service's main logic once the app is started
+        self.app.on_startup.append(self.run_service)
+
+        super().run(port=self.config["matchmaker"]["port"], *args, **kwargs)
+
+    async def run_service(self, app):
+        """Runs the service."""
+        await self.setup()
+        await self.bootstrap()
+
+        # Start service components
+        await asyncio.wait(
+            [self.new_champions_watcher_loop(), self.create_matches_loop()]
+        )
+
+    def get_champions_query(self):
+        """Returns a Django query selecting the champions in the tournament."""
+        all_chs = Champion.objects.filter(status="ready", deleted=False)
+
+        # TODO: move this configuration data to the Tournament model
+        deadline_str = self.config["matchmaker"].get("deadline", None)
+        include_staff = self.config["matchmaker"]["include_staff"]
+
+        if deadline_str:
+            deadline = datetime.datetime.fromisoformat(deadline_str)
+            all_chs = all_chs.filter(ts__lte=deadline)
+        if not include_staff:
+            all_chs = all_chs.filter(author__is_staff=False)
+
+        # Last champion of each user
+        # https://stackoverflow.com/questions/16074498
+        chs_ids = (
+            all_chs.values("author__id")
+            .annotate(max_id=Max("id"))
+            .values("max_id")
+        )
+
+        return Champion.objects.filter(pk__in=chs_ids)
+
+    async def setup(self) -> None:
+        """Configures MatchMaker."""
+        self.author, _ = get_user_model().objects.get_or_create(
+            username=self.config["matchmaker"]["author_username"]
+        )
+        logging.info(
+            "MatchMaker will use %s as author to schedule new matches",
+            self.author,
+        )
+
+        self.tournament, _ = Tournament.objects.get_or_create(
+            name=self.config["matchmaker"]["tournament_name"],
+            defaults={"author": self.author},
```

Comment:
What happens if the existing tournament author is not the one configured (`self.author`)?